src/sonardevices/OculusSonar.cpp

```diff
@@ -0,0 +1,555 @@
+                printf("OculusSonar: Received dummy message\n");
+                break;
+            default:
+                printf("OculusSonar: Received unknown message\n");
+                break;
+        }
+
+        // If there is additional data in buffer shift it to the front
+        memmove(rxBuff, &rxBuff[currPktSize], rxBuffPos - currPktSize);
+        rxBuffPos -= currPktSize;
+    
+    }
+
+    // Free memory
+    free(rxBuff);
+    rxBuff = nullptr;
+}
+
+void OculusSonar::processSimplePingResult(OculusMessages::OculusSimplePingResult *ospr){
+    uint8_t *startAddress = (uint8_t *)ospr;
+
+    uint32_t imageSize;
+    uint32_t imageOffset;
+    uint16_t beams;
+    uint16_t ranges;
+
+    uint16_t version = ospr->fireMessage.head.msgVersion;
+    
+    OculusMessages::OculusSimplePingResult2 *ospr2 = (OculusMessages::OculusSimplePingResult2 *)ospr;
+
+    // Check message version
+    switch (version){
+        case 2:
+            imageSize = ospr2->imageSize;
+            imageOffset = ospr2->imageOffset;
+            beams = ospr2->nBeams;
+            ranges = ospr2->nRanges;
+            rangeResolution = ospr2->rangeResolution;
+            break;
+        default:
+            imageSize = ospr->imageSize;
+            imageOffset = ospr->imageOffset;
+            beams = ospr->nBeams;
+            ranges = ospr->nRanges;
+            rangeResolution = ospr->rangeResolution;
+            break;
+    }
+    // Remember the beams and range bin count
+    this->beams = beams;
+    this->rangeBinCount = ranges;
+
+    // Check if image size is correct
+    if (ospr->fireMessage.head.payloadSize + sizeof(OculusMessages::OculusMessageHeader) == imageOffset + imageSize){
+        switch (version){
+            case 2:{
+                this->operatingFrequency = ospr2->frequency;
+                this->temperature = ospr2->temperature;
+                this->pressure = ospr2->pressure;
+                break;
+            }
+            default:{
+                this->operatingFrequency = ospr->frequency;
+                this->temperature = ospr->temperature;
+                this->pressure = ospr->pressure;
+                break;
+            }
+        }
+
+        uint32_t bearingTableSize = beams * sizeof(int16_t);
+        lastImage->bearingTable = std::make_unique<std::vector<int16_t>>(startAddress, startAddress + 122);
+
+        lastImage->imageHeight = ranges;
+        lastImage->imageWidth  = beams;
+        //lastImage->data = std::make_unique<std::vector<uint8_t>>(startAddress + imageOffset, startAddress + imageOffset + imageSize);
+        rosImageMessagePointer->data = std::vector<uint8_t>(startAddress + imageOffset, startAddress + imageOffset + imageSize);
+
+        // New image ready, notify all callbacks
+        SonarCallback cb;
+        std::lock_guard<std::mutex> lock(callbackMutex);
+        for (uint16_t i = 0; i < callbacks.size(); i++){
+            cb = callbacks.at(i);
+            cb(lastImage);
```

Comment:
@EulersBoiler You rendered the callback mechanism useless. The `lastImage` parameter does no longer contain an actual image. Please fix.